codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -77,6 +77,11 @@
     public static final Cumulator MERGE_CUMULATOR = new Cumulator() {
         @Override
         public ByteBuf cumulate(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf in) {
+            if (!cumulation.isReadable()) {
```

Comment:
might make sense to move this into `channelRead`?

Reply:
This patch specifically wants to permit the cumulator greater control over what buffer is saved to cumulation, in all circumstances.  e.g. in the case that a message aware cumulator knows that we are waiting for a small message, but the input buffer is very large, it would be wasteful to save the entire buffer.  Or, for instance, you may be able to have a single dedicated buffer for all instances of a handler on an eventLoop, avoiding any allocation/recycling in typical operation, only copying bytes when an incomplete message arrives - which if your messages are typically small means minimal wasted memory and copying.